### PR TITLE
[SPARK-50339][SPARK-50360][SS] Enable changelog to store lineage information

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

Break down https://github.com/apache/spark/pull/48355 into smaller PRs. 

## Changelog Reader / Writer
We purpose to save the lineage to the first line of the changelog files.

For changelog reader, there is an abstract function `readLineage` created. In `RocksDBCheckpointManager.getChangelogReader` function, the `readLineage` will be called right after the initialization of the changelog reader to update the file pointer to after the lineage. Subsequent `getNext` function won't be affecter because of this.

For changelog writer, there is an abstract function `writeLineage` that writes the lineage. This function will be called before any actual changelog data is written in `RocksDB.load()`.

The lineage is stored as json.

### Why are the changes needed?

Continue development of SPARK-49374

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added unit test

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (siying) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:299]: Can we represent lineage item in a case class rather than a tuple? It's hard to add a field (e.g. hasSnapshot) later with a tuple. It's also easier to read.

The way of lazily reading lineage is not robust to me. There is an implicit contract that `lineage` has to be referenced before `getNext()`. It will be better if the code doesn't build on implicit contract like this. It will be slightly better if initialization is a separate function call. Or maybe even crate a general `init()` function that does lineage reading.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (siying) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:214]: I guess here we reference a variable to force it to be initialized? I think it would be easier to understand if we explicitly create a function such as `loadLineage()` or something like that.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (siying) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:777]: Should we cover both V1 and V2? I saw lineage is written in both.

**Review by siying — APPROVED** [review]
Generally LGTM. I left a test comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:118]: Is there anything that guarantees that this is going to be called first so that it goes into the first line?

**Review by brkyvz — APPROVED** [review]
I am not a fan of the interface. You're essentially requiring the creator of the Changelog writer to first call `writeLineage`, which feels like an abstraction leak and makes it error prone. Can the lineage be passed in to the writer instead instead of just the `enableStateStoreCheckpointIds`? Or require that `writeLineage` was called by an internal flag before put/merge/delete is called?

I would rather still prefer the former than the latter though

**Comment by HeartSaVioR** [comment]
+1 to @brkyvz about the interface and I also prefer former on the suggestion.

Also, I'd like to see the changelog file to be "self described". It is indeed a change on the storage format to introduce the lineage. For the change of storage format, I'd rather say, we need to bump the version. I understand changelog file will be accessed after looking into commit log hence it's clear whether the query has uuid to be available or not, but still, we are drawing another axis of the versioning for the storage format of the changelog file. I'd like to see this be flattened (single axis of the timeline for the versioning).

version 2 is superior to version 1 - it has more overhead but cover more functionality and I don't think the overhead would be outstanding. That said, why not come up with version 3 based on version 2 + lineage? I'm even OK with version 3 for version 1 + lineage and version 4 for version 2 + lineage if we are not very sure about the overhead.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:157]: one parameter per line please. Also please add scaladoc

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:117]: keep it protected?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:282]: Seems like this is missing `writeVersion` since it extends V1 and not V2...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:302]: this becomes tricky to guarantee based on class/trait/abstraction initialization...

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:460]: this is weird...

**Review by brkyvz — CHANGES_REQUESTED** [review]
This V3, V4 stuff is really error prone... Can we please just introduce V3? Then introduce a `writeHeader` to `StateStoreChangelogWriter`. V1 doesn't write a version, V2 writes a version, V3 writes a version and lineage in this method.

Then, you should only have one Reader class. It just reads the first line, from there it knows whether it is v1 or something else. If it is a version it doesn't understand, then it'll throw an error along the lines of: "I don't know how to read this file". It can figure out the lineage as well if it is V3. Then it can read the data. The problem right now is that you've totally broken forward compatibility of changelogs - rather V1 and V2 is going to just fail with a FileNotFoundException if someone upgrades and then rolls back their Spark rather than a better error message of, "I can't read this stream"

**Comment by WweiL** [comment]
> just introduce V3?
This is sadly not possible imho. Because v1 and v2 not only differ in whether having a "v2" in the first line, but their format are different. We likely still need to have v3 and v4.

> Then, you should only have one Reader class. It just reads the first line, from there it knows whether it is v1 or something else.

Similarly for this one, I think a way out is to have a helper class that does this, i.e. it reads the version from the actual changelog file. Then the reader version should be decided by it. This is a tech debt indeed, I filed a spark ticket for it: https://github.com/WweiL/oss-spark/blob/8e439fcc43513f3048b51a8422c0ec136a0b624d/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala#L211-L215 Do you think we can resolve it in a separate PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:277]: Scaladoc please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:292]: scaladoc please. Why is this a new version? How is the encoding different

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:337]: you might want to turn this into a better error message saying that this is probably coming from a future Spark version. Please upgrade your Spark and use the Error Code framework to do so

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:341]: this can remain an assert

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:421]: Can you have this be called in `StateStoreChangelogReader` initialization? I'd have the following in `StateStoreChangelogReader`
```scala
protected def readVersion(): String = input.readUTF()

final def verifyVersion(): Unit = {
  if(!readVersion.stripPrefix("v") == version.toString) {
    throw ...
  }
}

verifyVersion()
 ...
}

class StateStoreChangelogReaderV1 {
  override protected def readVersion(): String = "v1"
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:462]: don't call these individually in all the formats... Error prone for future versions

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:421]: Also if `readVersion` fails, then you also probably want to throw the "Upgrade Spark" error

**Review by brkyvz — CHANGES_REQUESTED** [review]
I don't like the fact that we didn't initially implement these files with a proper protocol version, but at least we can do an assertion on the version checks...

Left some recommendations and please add tests for all the failure cases too on the versioning

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:292]: There is a `merge ` operation in V2 reader / writer, also the eof token is different

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:421]: Yes this is the ideal way, but the problem is in v1 there is no version written in the first line. Hence all the problems. If written in this way `verifyVersion` would always throw in v1 changelogs

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:462]: Yes I agree.. but for the same reason above (only v1 does not have version written in the first line), we need to handle this in a case by case fashion

**Comment by WweiL** [comment]
I ended up adding the helper class `StateStoreChangelogReaderFactory` to construct the state reader. PTAL @siying @HeartSaVioR @brkyvz

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:496]: that's why I had overridden `readVersion` above in `StateStoreChangelogReaderV1`

**Review by brkyvz — CHANGES_REQUESTED** [review]
I think you're leaking streams in StateStoreChangelogReaderFactory. Can I ask you for a few more tests:

 - Read a new version (v12) and make sure the error gets thrown
 - Accidentally read a v1 file with v3 configurations. See what happens
 - In the afterEach of tests, make sure that there are no leaked input/output streams

Will a stream be able to upgrade from 1 -> 3 and 2 -> 4? It seems impossible because the file names are not going to match

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:496]: I ended up keeping this. Because `StateStoreChangelogReaderV3` extends `StateStoreChangelogReaderV1` (for it's iterator functions), now the readVersion is also inherited. We could override it by 
```
class StateStoreChangelogReaderV3 {
  override protected def readVersion(): String = input.readUTF()
}
```

Also future new readers extending this V1 again needs to override `input.readUTF()`. I'm thinking this defeats the purpose of overrding the readVersion override in V1, which is try to prevent unintentional overrides. 

Sadly because of this version difference it looks inevitable that there is no way to perfectly prevent unintentional overrides..

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:111]: existing: why are these `var` instead of `val`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:296]: remove "starting from version 3"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:348]: maybe make this a `lazy val` so that the stream is only created once it is accessed

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:354]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (brkyvz) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:341]: nit: space before `{`

**Review by brkyvz — APPROVED** [review]
LGTM. thanks for adding all the tests

**Comment by brkyvz** [comment]
@HeartSaVioR can I again ask for your help in merging this once the final comments are addressed?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (WweiL) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:111]: in abort() compressedStream is set to null

**Comment by HeartSaVioR** [comment]
Same thing; I'm merging on behalf of @brkyvz .

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala
Comment (tedyu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala:391]: I think `input` should be set to `null` in this block.

**Comment by tedyu** [comment]
A factory's responsibility should be the creation of objects. Adding constraints on how the factory can be used (e.g., only once for `constructChangelogReader`) is a violation of the Single Responsibility Principle (SRP).

The single use violates the expectations of a reusable factory. It creates confusion for clients of the factory, as factories are typically expected to produce multiple instances when invoked repeatedly.

**Comment by WweiL** [comment]
@tedyu So a better way of designing is to put the class parameter to `constructChangelogReader`? I'd be happy to approve if there are followups filled : )

**Pull request closed without merging**
